### DISCUSSION TITLE
Export as OpenAPI

### DISCUSSION THREAD

**Discussion by ecarter202** [Kong/insomnia#3737] [Ideas]
Would like to be able to export as OpenAPI. Pretty simple. Thanks.

**Comment by welcome** [comment]
👋 Thanks for opening your first issue! If you're reporting a 🐞 bug, please make sure
you include steps to reproduce it. If you're requesting a feature 🎁, please provide real
use cases that would benefit. 👪

To help make this a smooth process, please be sure you have first read the
[contributing guidelines](https://github.com/getinsomnia/insomnia/blob/master/CONTRIBUTING.md).

**Comment by mtsmachado8** [comment]
It would be great! please consider it, thumbs up!

**Comment by gschier** [comment]
This would be awesome, I agree!

**Comment by dingyaguang117** [comment]
really need this feature

**Comment by pete001** [comment]
Yeah this would be epic

**Comment by reilg** [comment]
This would be incredibly useful.

**Comment by gruberdev** [comment]
You can actually do it your own in a fairly easy way (especially if you rely on a CI), whenever you sync your API to a remote Git, it creates a `.insomnia`  folder, where inside it automatically creates an `ApiSpec` folder, inside there's one `.yaml` file for each one of your API specs. 
The good news is: each `.yaml` is elaborated in such a consistent way that if you exclude the first 3 lines and the last 5 lines of every `.yaml` file inside `.insomnia/ApiSpec/`, you'll obtain an OpenAPI compatible file. 

I think a simple JS editor can be done meanwhile, but I rather see that implemented as an actual feature...

**Comment by ghostbody** [comment]
That will be awesome while [swaggymnia](https://github.com/mlabouardy/swaggymnia)v is a decent solution currently.

**Comment by sanchitsharma** [comment]
> That will be awesome while [swaggymnia](https://github.com/mlabouardy/swaggymnia)v is a decent solution currently.

I tried swaggymnia but it is not working with Insomnia v4 and the latest insomnia is only allowing to export in Insomnia v4. Any idea how to proceed?

**Comment by ghostbody** [comment]
@sanchitsharma try: https://github.com/Fyb3roptik/swaggomnia

**Comment by develohpanda** [comment]
> The good news is: each `.yaml` is elaborated in such a consistent way that if you exclude the first 3 lines and the last 5 lines of every `.yaml` file inside `.insomnia/ApiSpec/`, you'll obtain an OpenAPI compatible file.

While Insomnia currently doesn't _generate_ an OpenAPI spec (and I agree it would be a useful feature), just chiming in that we've recently released [`inso`](https://github.com/Kong/insomnia/tree/develop/packages/insomnia-inso#-inso-export-spec-identifier), a companion CLI through which you can export the spec written in designer into a file (as an alternative to the comment quoted). 😄 

`inso export spec specName -o output.yaml`

**Comment by codeasashu** [comment]
I've created a very basic [npm package](https://www.npmjs.com/package/insomnia-openapi-converter/) here which may suit the needs. Can someone give it a try. Its an early release and thus not a cli usable package, but it does borrow some of the items from swaggymnia

https://www.npmjs.com/package/insomnia-openapi-converter/

**Comment by NickAcPT** [comment]
The fragmentation on this project is a bit worrying and confusing.

With that out of the way, I'm making it clear: I'd love to have this included itself on the Insomnia Core application. Other than the (currently unusable) npm package mentioned above, has there been any progress on this?

**Comment by develohpanda** [comment]
Hi @NickAcPT - there was some confusion with the comments/questions and the original feature request here, so I cleared that in my comment above. However, the original feature request currently does not have an progress.

I believe there are enough plugin APIs available in order to create a plugin for Insomnia Core to be able to achieve this.

**Comment by HeathNaylor** [comment]
Is there any progress on this? I would expect to see OpenAPI as a drop down option when exporting from Insomnia. This would be a huge win over Postman currently. Right now if I want to even get close to generating library code from a non-openapi format (Postman Collection or Insomnia) I have to use something like `apimatic.io` to transform it, no other tools seem to work.

**Comment by develohpanda** [comment]
Hi @HeathNaylor, unfortunately no progress yet but it may be possible to create a plugin to use `apimatic.io` in the meantime 👀

Adding this natively _is_ on the roadmap but not in the immediate future as far as I know.

**Comment by codeasashu** [comment]
Hi @NickAcPT I am sorry the plugin didn't worked out for you. ~Let me fix it and get back~ I've fixed the plugin and added a test for very basic functionality: https://www.npmjs.com/package/insomnia-oas-converter

If you can, please file an issue so I can continue making package better.

**Comment by abitrolly** [comment]
Why it was moved from https://github.com/Kong/insomnia/issues/1778#event-4962582884 into this discussion? What wrong with the ticket?

**Reply by wdawson** [reply]
Hi @abitrolly , there's nothing wrong with the ticket. We're just re-organizing any feature requests and ideas to GitHub discussions so that we can better talk about these. We'll still be active here and consider these the same as before 👍

**Comment by sabat24** [comment]
So basically how should I use that converter directly in Insomnia or via command line? Most of converters are fired directly from CLI using node. However I can't see such a possibility here. There is a statement that:
> you can use this as a plugin

But how can I do this?

**Reply by develohpanda** [reply]
Sorry I'm not sure I follow 🤔 we don't currently have plugins (that I know of) for automatic conversion to OpenAPI. I'm sure https://www.npmjs.com/package/insomnia-oas-converter could be packaged into a plugin but right now it looks like you'll have to use it through JS using Node.

**Reply by codeasashu** [reply]
@sabat24 In insomnia, You can easily create your own plugins. Follow the steps below to use the [insomnia-oas-converter](https://www.npmjs.com/package/insomnia-oas-converter) in your own plugin:

1. You can also easily get the insomnia workspace data in your plugin (see [Workspace Actions](https://support.insomnia.rest/article/173-plugins#workspace-actions)).
2. Install [insomnia-oas-converter](https://www.npmjs.com/package/insomnia-oas-converter ) as a dependency in your plugin, and feed the output from step 1 into this. You'll get the OAS data as the exported output. See the docs for the desired format.
3. Export the OAS data in your choosing.
4. Use bundler (eg. webpack) to compile the plugin and use with insomnia

Hope that helps you get started. Should you feel any issues, feel free to open it in the [repo](https://github.com/codeasashu/insomnia-oas-converter) and I'll try to improve the documentation.

**Comment by mehmetaydogduu** [comment]
apimatic.io works well.
